### PR TITLE
`URL.build()`: Raise `TypeError` if `host` argument is `None`

### PR DESCRIPTION
**Pull request by paulpapacz**
## What do these changes do?

Make `URL.build()` raise a `TypeError` if `host` argument is `None`

## Are there changes in behavior for the user?

No.

## Related issue number

#808

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

File: tests/test_url_build.py [hunk 1]
```diff
@@ -239,6 +239,11 @@ def test_build_with_authority_with_path_without_leading_slash():
         URL.build(scheme="http", host="example.com", path="path_without_leading_slash")
 
 
+def test_build_with_none_host():
+    with pytest.raises(TypeError):
```
Comment (webknjaz) [inline tests/test_url_build.py:243]: It is always important to use the `match` argument with `pytest.raises()`.

**Review by webknjaz — COMMENTED** [review]
FWIW I there are valid cases for not having a hostname. For example, this is how one would represent local filesystem paths or URLs, or UNIX file socket URLs, or abstract socket URLs. Hostnames are needed where a URI would represent an IP-network-accessible resource but a generic case would include things like `http+unix:///var/tmp/aiohttp-server.sock` even for web apps that are intended to operate locally. YARL is a generic lib so we shouldn't ignore use-cases that exist outside web. There's `file:///etc/hosts` or `git+ssh://../another-repo/.git` that are valid URLs. And there are URLs that are used in protocols targeting `xdg-open`/MIME thingies. Like `steam://some-app?params` or `twitter://username` etc.

If we want to force people pass a non-`None` host, we should at least make sure there are tests for these cases allowing the end-users to construct corresponding URLs with something like an empty string, for example.

**Comment by webknjaz** [comment]
But I guess having a `None` value representing “no hostname” seems to be better than an empty string that might have a different semantic meaning.
@Dreamsorcerer WDYT?

File: tests/test_url_build.py (hunk 1 above)
Comment (paulpapacz) [reply tests/test_url_build.py]: Added a `match` now but none of the other test cases uses it (just FYI).

Comment (webknjaz) [reply tests/test_url_build.py]: That's unfortunate. And a chance to contribute more refactoring PRs for someone.

File: CHANGES/808.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Made ``URL.build()`` raise a ``TypeError`` if the ``host`` argument is ``None``.
```
Comment (webknjaz) [inline CHANGES/808.bugfix.rst:1]: It should be possible to link these, thanks to intersphinx:
```suggestion
Made :py:meth:`URL.build` raise a :py:exc:`TypeError` if the ``host`` argument is :py:data:`None` — by :user:`paulpapacz`.
```

**Comment by webknjaz** [comment]
Oh, and one other case is relative URIs which only need path+qs.

**Comment by Dreamsorcerer** [comment]
> But I guess having a `None` value representing “no hostname” seems to be better than an empty string that might have a different semantic meaning. @Dreamsorcerer WDYT?

Looks like the current situation is that it's not designed for it. At a glance, it looks much the same as the other parameters in that block, and fails if you try to use `None`:
```
>>> yarl.URL.build(host=None)  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python3.10/site-packages/yarl/_url.py", line 246, in build
    cls._validate_authority_uri_abs_path(host=host, path=path)
  File "/home/ubuntu/.local/lib/python3.10/site-packages/yarl/_url.py", line 699, in _validate_authority_uri_abs_path
    if len(host) > 0 and len(path) > 0 and not path.startswith("/"):
TypeError: object of type 'NoneType' has no len()
```

I don't think there's any meaningful different to `""` or `None` representing "no hostname", the result would surely look the same, but would require extra code to convert the `None`.

So, this change looks fine to me. We should get type annotations added though, which would make it more explicit.

**Comment by Dreamsorcerer** [comment]
Nevermind, it use stub files. This is already typed as `str`: https://github.com/aio-libs/yarl/blob/master/yarl/__init__.pyi#L59

**Review by webknjaz — APPROVED** [review]
@Dreamsorcerer I'll leave the final merge to you if you don't mind.

**Pull request merged**